src/program.cpp

```diff
@@ -105,829 +63,148 @@ program::program(const program& p) { assign(p); }
 
 shape program::get_parameter_shape(std::string name) const
 {
-    auto ins = std::find_if(
-        impl->instructions.begin(), impl->instructions.end(), [&](const instruction& x) {
-            if(x.name() == "@param")
-            {
-                return any_cast<builtin::param>(x.get_operator()).parameter == name;
-            }
-            else
-            {
-                return false;
-            }
-        });
-    if(ins != this->end())
-        return ins->get_shape();
-    else
-        return {};
+    return main_module.get_parameter_shape(std::move(name));
 }
 
 std::vector<std::string> program::get_parameter_names() const
 {
-    std::vector<std::string> result = impl->input_names;
-    std::unordered_set<std::string> params;
-    for(auto&& ins : impl->instructions)
-    {
-        if(ins.name() == "@param")
-        {
-            auto&& name = any_cast<builtin::param>(ins.get_operator()).parameter;
-            params.insert(name);
-        }
-    }
-    erase_if(result, [&](auto&& name) { return params.count(name) == 0; });
-    return result;
+    return main_module.get_parameter_names();
 }
 
 instruction_ref program::get_parameter(std::string name) const
 {
-    auto ins = std::find_if(
-        impl->instructions.begin(), impl->instructions.end(), [&](const instruction& x) {
-            if(x.name() == "@param")
-            {
-                return any_cast<builtin::param>(x.get_operator()).parameter == name;
-            }
-            else
-            {
-                return false;
-            }
-        });
-    if(ins != this->end())
-        return ins;
-    else
-        return this->end();
+    return main_module.get_parameter(std::move(name));
 }
 
 std::unordered_map<std::string, shape> program::get_parameter_shapes() const
 {
-    std::unordered_map<std::string, shape> result;
-    for(auto&& ins : impl->instructions)
-    {
-        if(ins.name() == "@param")
-        {
-            auto&& name  = any_cast<builtin::param>(ins.get_operator()).parameter;
-            result[name] = ins.get_shape();
-        }
-    }
-    return result;
+    return main_module.get_parameter_shapes();
 }
 
 bool program::has_instruction(instruction_ref ins) const
 {
-    return std::find_if(
-               impl->instructions.begin(), impl->instructions.end(), [&](const instruction& x) {
-                   return std::addressof(*ins) == std::addressof(x);
-               }) != impl->instructions.end();
+    return main_module.has_instruction(ins);
 }
 
-std::size_t program::size() const { return impl->instructions.size(); }
-instruction_ref program::begin() const { return impl->instructions.begin(); }
-instruction_ref program::end() const { return impl->instructions.end(); }
+std::size_t program::size() const { return main_module.size(); }
+instruction_ref program::begin() const { return main_module.begin(); }
+instruction_ref program::end() const { return main_module.end(); }
```

Comment:
This should iterate over the modules. Since we only have one module it can be written like this:

```cpp
module* program::begin() const { return &main_module; }
module* program::end() const { return begin()+1; }
```

Reply:
Should we also add an API for the instruction begin() and end() of a program (it is actually the begin() and end instructions of the main module)?